### PR TITLE
ignition-fire-up

### PR DESCRIPTION
**Pull request by calvix**
Migrate mayu to ignition
towards: https://github.com/giantswarm/giantswarm/issues/2121

changes:
* migrated cloudconfig to ignition
* added filesManager (ignition does not support files in the ignition itself so we put them in separate files in dir `./files` and then load them via fileManager,  render the template values and encode with base64 into ignition)
* ignition templates are in yaml but result ignition is json (yaml is easier to read)
* removed 2 phase installation
* removed dependency on `mayu-infopusher`, serial is fetched via ipxe script
* tweaking hostname from random hash to machine ip (we can the use for better identification in k8s)
there are also few changes to g8s in hive configs - https://github.com/giantswarm/hive/pull/193

 I wanted to separate this changes with cleaning mayu from cloudconfig so its not too big PR.
Next, follow up will remove all unused code and cloudconfig support.

I need to update the docs for sure as there are few new things. But this make sense  when the cloudconfig stuff is removed.

### CODE REVIEW

File: pxemgr/ignition.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 	"io"
 	"reflect"
 
-	"github.com/coreos/ignition/config/types"
+	"github.com/coreos/ignition/config/v2_0/types"
```
Comment (calvix) [inline pxemgr/ignition.go]: locking to ignition 2.0 as 2.1 is experimental/alpha and there are few breaking changes

File: main.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ const (
 	DefaultIgnitionTemplateSnippets string = "./template_snippets/ignition"
 	DefaultDNSMasq                  string = "/usr/sbin/dnsmasq"
 	DefaultImagesCacheDir           string = "./images"
+	DefaultFilesDir                 string = "./files"
```
Comment (calvix) [inline main.go]: this is the new files dir, in mayu its mounted similarly like `templates` and `template_snippets`

File: pxemgr/filemanager.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+		for _, file := range fileList {
+			tmpl, err := template.ParseFiles(path.Join(mgr.filesDir, dir.Name(), file.Name()))
+			if err != nil {
+				glog.Errorf("Failed to file: %s, error: %#v", path.Join(mgr.filesDir, dir.Name(), file.Name()), err)
+			}
+			var data bytes.Buffer
+			tmpl.Execute(&data, ctx)
+
+			files[dir.Name()+"/"+file.Name()] = base64.StdEncoding.EncodeToString(data.Bytes())
```
Comment (calvix) [inline pxemgr/filemanager.go]: the file is stored in `map[string]string` where a key to the file is the path ie: `conf/os-hardning.conf`  (files are atm in the hive repo PR)

File: pxemgr/ipxe_handlers.go [hunk 4]
```diff
@@ -30,18 +30,46 @@ const (
 	qemuKernelFile   = "coreos_production_qemu.vmlinuz"
 
 	defaultProfileName = "default"
+
+	kvmStaticSerial = "0123456789"
```
Comment (calvix) [inline pxemgr/ipxe_handlers.go]: this is hack for kvm running on gridscale/KVM as the serial is static and has this value, if we encounter it we use UUID for server serial instead

File: hostmgr/cluster.go [hunk 5]
```diff
@@ -149,7 +150,10 @@ func (c *Cluster) CreateNewHost(serial string) (*Host, error) {
 	newHost.Cabinet = cabinet
 	newHost.MachineOnCabinet = machineOnCabinet
 	newHost.MachineID = machineID
-	newHost.Hostname = machineID[:16]
+	if newHost.InternalAddr != nil {
```
Comment (r7vme) [inline hostmgr/cluster.go]: What if InternalAddr is nil?

File: pxemgr/filemanager.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	"encoding/base64"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"path"
+	"text/template"
+)
+
+type Files map[string]string
+
+func (mgr *pxeManagerT) RenderFiles(ctx interface{}) *Files {
```
Comment (r7vme) [inline pxemgr/filemanager.go]: Please execute `gofmt`. Ideally we also need `golint`, but i'm ok to do it separately

Comment (calvix) [reply pxemgr/filemanager.go]: `gofmt` is needed as CI won't succeed unless code is `gofmt`ed
what looks funky? the empty lines?

File: hostmgr/cluster.go (hunk 5 above)
Comment (calvix) [reply hostmgr/cluster.go]: the thing is that this function is run few times, and in each run it  adds some extra info ( I am not sure why it works this ways, its kinda legacy, maybe we can later improve)

File: main.go (hunk 2 above)
Comment (teemow) [reply main.go]: what about the old templates then? can you also remove `DefaultFirstStageScript` and `DefaultLastStageCloudconfig`?

Comment (calvix) [reply main.go]: `files` dir is there for files not for templates so we need both, but yes I will be removing all the stuff like `DefaultFirstStageScript` in the clean PR which will follow after this one.

File: pxemgr/config.go [hunk 7]
```diff
@@ -53,9 +53,10 @@ type Profile struct {
 }
 
 type Network struct {
-	Interface      string
-	BindAddr       string `yaml:"bind_addr"`
-	BootstrapRange struct {
+	Interface        string
+	MachineInterface string `yaml:"machine_interface"`
```
Comment (corest) [inline pxemgr/config.go]: Wanted to find this :heart:

Comment (calvix) [reply pxemgr/config.go]: I think it makes sense to rename the `interface` to `pxe_interface`  for easier understanding.

File: pxemgr/ipxe_handlers.go (hunk 4 above)
Comment (teemow) [reply pxemgr/ipxe_handlers.go]: would be good to put such comments in the code itself :)

**Pull request merged**